### PR TITLE
sd-shutdown improvements

### LINKED ISSUE #8155: shutdown[1]: Failed to wait for process: Protocol error
### Submission type

 - Bug report

### systemd version the issue has been seen with
237.0-2

### Used distribution
archlinux 64 bits

### In case of bug report: Expected behaviour you didn't see

> no error message should be seen at extinction of the PC

### In case of bug report: Unexpected behaviour you saw

> I see the message : 
shutdown[1]: Failed to wait for process: Protocol error

### In case of bug report: Steps to reproduce the problem
- use archlinux 64 bits
- install the last archlinux package of systemd ( 237.0-2 )
- during the shutdown you will see the message : 
shutdown[1]: Failed to wait for process: Protocol error

see the attached file ( screen capture )

### PR DESCRIPTION
**Pull request by medhefgo**
This should make sure the shutdown process only logs when it should. I'm not too sure about the last patch, though. It feels a bit ugly, but it gets the job done.

I am also wondering whether the sd-shutdown umount loop shouldn't be turned into a while loop. We are already checking if we are making any changes during iterations, and currently, there is potential for mounts not to get handled should we have a crazy nested setup that goes beyond FINALIZE_ATTEMPTS.

Fixes #8155

### CODE REVIEW

File: src/core/umount.c [hunk 1]
```diff
@@ -604,25 +610,25 @@ static int dm_points_list_detach(MountPoint **head, bool *changed) {
...
 
         LIST_FOREACH_SAFE(mount_point, m, n, *head) {
 
-                if (major(rootdev) != 0)
+                if (major(rootdev) != 0) {
                         if (rootdev == m->devnum) {
                                 n_failed ++;
                                 continue;
                         }
+                }
```
Comment (keszybz) [inline src/core/umount.c]: I don't think there's any reason for this chunk.

File: src/core/umount.c [hunk 2]
```diff
@@ -146,8 +149,28 @@ static int mount_points_list_get(MountPoint **head) {
...
+                        /* mount(2) states that mount flags and options need to be exactly the same
+                         * as they were when the filesystem was mounted, except for the desired
+                         * changes. So we reconstruct both here and adjust them for the later
+                         * remount call too. */
+
+                        r = mount_option_mangle(flags, 0, &m->remount_flags, &unknown_flags);
+                        if (r < 0)
+                                return r;
+                        if (!isempty(unknown_flags))
+                                log_warning("Ignoring unkown mount flags '%s'.", unknown_flags);
```
Comment (keszybz) [inline src/core/umount.c]: Typo in "unkown"

File: src/core/umount.c [hunk 3]
```diff
@@ -146,8 +149,28 @@ static int mount_points_list_get(MountPoint **head) {
...
+
+                        r = mount_option_mangle(options, m->remount_flags, &m->remount_flags, &m->remount_options);
+                        if (r < 0)
+                                return r;
+
+                        /* MS_BIND is special. If it is provided it will only make the mount-point
+                         * read-only. If left out, the super block itself is remounted, which we want. */
+                        m->remount_flags = (m->remount_flags|MS_REMOUNT|MS_RDONLY) & ~MS_BIND;
```
Comment (keszybz) [inline src/core/umount.c:174]: For the whole commit: please provide justification for this change in the patch comment. What are the benefits? For which file systems does this matter?

File: src/core/umount.c [hunk 4]
```diff
@@ -519,16 +519,19 @@ static int mount_points_list_umount(MountPoint **head, bool *changed) {
...
                          *
                          * Since the remount can hang in the instance of
                          * remote filesystems, we remount asynchronously
-                         * and skip the subsequent umount if it fails */
+                         * and skip the subsequent umount if it fails. */
                         if (remount_with_timeout(m) < 0) {
-                                if (nonunmountable_path(m->path))
+                                /* Remount failed, but try unmounting anyway,
+                                 * unless this is a mount point we want to skip. */
+                                if (nonunmountable_path(m->path)) {
```
Comment (keszybz) [inline src/core/umount.c]: Hm, isn't this patch made obsolete by the subsequent patch to skip remount-ro on api filesystems?

**Comment by medhefgo** [comment]
Made some changes.

Originally, I thought that the kernel might be complaining about the flags not being the same (sans the ro) when remounting the cgroup tree. But it turned out not to do anything. I decided to keep it, because the manpage said that the flags should be set that way after all. In fact, I was the one that added the code to reuse the mount options some time ago. I simply must've skipped this detail the last time, or I would've added it then.

Yes, not remounting api filesystems in the first place would make the umount after failed remount obsolete. But I figured it would future-proof the whole thing. If some new (api) fs should come up that cannot be remounted but also isn't detected as api yet, we would once again be in a situation where we cannot unmount / or /oldroot just because of some (api) fs mounted below.

**Comment by keszybz** [comment]
OK, this seems good to merge. This area is tricky, so I'm not 100% sure that it's all correct, but we're pretty far from a release now, so if there are any regressions we should catch them. Let's keep an eye out.

**Comment by sueridgepipe** [comment]
Has this been tested with a LUKS encrypted root partition?

This patch was just introduced into Manjaro unstable branch.  All my systems have an encrypted root partition and all crash during shutdown / reboot.  Hard power off is required.

screenshot_20180314_084333-1

This is a virtualbox screen grab, but the text is identical on bare metal systems with the same crash.  There is obviously more text but not able to scroll up.

The following systemd-cryptsetup errors are also generated during shutdown / reboot.
```
Mar 13 manjaro systemd-cryptsetup[14160]: Failed to deactivate: Device or resource busy
Mar 13 manjaro systemd-cryptsetup[14161]: Failed to deactivate: Device or resource busy
Mar 13 manjaro kernel: watchdog: watchdog0: watchdog did not stop!
```
With a non encrypted root partition reboot and shutdown work normally.

**Comment by medhefgo** [comment]
I haven't tested this with encrypted root. Can you git bisect this to a specific commit?

**Comment by philmmanjaro** [comment]
@sueridgepipe: please get the sources of our package and add the commits one by one after [line 67](https://github.com/manjaro/packages-core/blob/master/systemd/PKGBUILD#L67) to see which commit is the culprit.

**Comment by keszybz** [comment]
I think I have a fix for you, no need to bisect. I'll push a PR shortly.

**Comment by keszybz** [comment]
See #8425 and in particular https://github.com/systemd/systemd/pull/8452/commits/b93618644bf24bc0cd87adb1346955d783f6f9b8.

**Comment by philmmanjaro** [comment]
@keszybz: I assume you mean pull #8452

**Comment by evverx** [comment]
I added the "ci-fails" and "merged-but-needs-fixing" labels because the CI failed with
```sh
[[0;32m  OK  [0m] Stopped Remount Root and Kernel File Systems.
[[0;32m  OK  [0m] Removed slice system-systemd\x2dfsck.slice.
[[0;32m  OK  [0m] Stopped Cryptography Setup for varcrypt.
[[0;32m  OK  [0m] Reached target Unmount All Filesystems.
[[0;32m  OK  [0m] Removed slice system-systemd\x2dcryptsetup.slice.
[[0;32m  OK  [0m] Reached target Shutdown.
[[0;32m  OK  [0m] Reached target Final Step.
         Starting Power-Off...
[   21.831426] Kernel panic - not syncing: Attempted to kill init! exitcode=0x0000008b
[   21.831426] 
[   21.832309] CPU: 0 PID: 1 Comm: systemd-shutdow Not tainted 4.4.0-116-generic #140-Ubuntu
[   21.832309] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS Ubuntu-1.8.2-1ubuntu1 04/01/2014
[   21.832309]  0000000000000086 609837e11178841e ffff88001c73fc48 ffffffff813ffc13
[   21.832309]  ffffffff81cbc7b0 ffff88001c73fce0 ffff88001c73fcd0 ffffffff81191327
[   21.832309]  ffff880000000010 ffff88001c73fce0 ffff88001c73fc78 609837e11178841e
[   21.832309] Call Trace:
[   21.832309]  [<ffffffff813ffc13>] dump_stack+0x63/0x90
[   21.832309]  [<ffffffff81191327>] panic+0xd3/0x227
[   21.832309]  [<ffffffff810865c4>] do_exit+0xaf4/0xb00
[   21.832309]  [<ffffffff81086653>] do_group_exit+0x43/0xb0
[   21.832309]  [<ffffffff81092e74>] get_signal+0x294/0x600
[   21.832309]  [<ffffffff8102e567>] do_signal+0x37/0x6f0
[   21.832309]  [<ffffffff8109073a>] ? signal_wake_up_state+0x2a/0x30
[   21.832309]  [<ffffffff81090876>] ? complete_signal+0x106/0x1f0
[   21.832309]  [<ffffffff810cd681>] ? __raw_callee_save___pv_queued_spin_unlock+0x11/0x20
[   21.832309]  [<ffffffff8100320c>] exit_to_usermode_loop+0x8c/0xd0
[   21.832309]  [<ffffffff81003c26>] prepare_exit_to_usermode+0x26/0x30
[   21.832309]  [<ffffffff8184fbe2>] retint_user+0x8/0x50
[   21.832309] Kernel Offset: disabled
[   21.832309] ---[ end Kernel panic - not syncing: Attempted to kill init! exitcode=0x0000008b
[   21.832309] 
```

**Pull request merged**